pkg/istio-agent/xds_proxy.go

```diff
@@ -123,54 +128,97 @@ func initXdsProxy(ia *Agent) (*XdsProxy, error) {
 	if err = proxy.initCertificateWatches(ia, proxy.stopChan); err != nil {
 		return nil, err
 	}
+
+	go proxy.healthChecker.PerformApplicationHealthCheck(func(healthEvent *health.ProbeEvent) {
```

Comment:
Curious - What is the upside to using a callback as opposed to directly using a channel here?

Reply:
I was googling the same thing when I wrote this. I found https://www.jtolio.com/2016/03/go-channels-are-bad-and-you-should-feel-bad/ - see "Callbacks are strictly more powerful and don’t require unnecessary goroutines" - which I agree with. Basically, if you use a channel you are forcing the caller to use a certain pattern. callback is more flexible

Reply:
good point. It also offers flexibility in tests from what it looks like.